src/generic/eigen_solver.cc

```diff
@@ -700,19 +736,205 @@ namespace oomph
+    solve_eigenproblem_helper(
+      problem_pt, n_eval, alpha_eval, beta_eval, eigenvector_aux);
+
+    // Now move auxiliary data into actual complex eigenvalues
+    // Instrutions from lapack qz (where VR translates into eigenvector_aux):
+    //        VR is DOUBLE PRECISION array, dimension (LDVR,N)
+    //        If JOBVR = 'V', the right eigenvectors v(j) are stored one
+    //        after another in the columns of VR, in the same order as
+    //        their eigenvalues. If the j-th eigenvalue is real, then
+    //        v(j) = VR(:,j), the j-th column of VR. If the j-th and
+    //        (j+1)-th eigenvalues form a complex conjugate pair, then
+    //        v(j) = VR(:,j)+i*VR(:,j+1) and v(j+1) = VR(:,j)-i*VR(:,j+1).
+    //        Each eigenvector is scaled so the largest component has
+    //        abs(real part)+abs(imag. part)=1.
+    unsigned n = problem_pt->ndof();
+    eigenvector.resize(n);
+    unsigned eval_count = 0;
+    while (eval_count < n)
+    {
+      // i-th eigenvalue is real:
+      if (alpha_eval[eval_count].imag() == 0.0)
+      {
+        // Resize the single eigenvector associated with this
+        // single real eigenvalue
+        eigenvector[eval_count].resize(n);
+        for (unsigned j = 0; j < n; ++j)
+        {
+          eigenvector[eval_count][j] =
+            std::complex<double>(eigenvector_aux[j][eval_count], 0.0);
+        }
+        eval_count++;
+      }
+      // Assume (and check!) that complex conjugate pairs follow each other
+      // as implied by
+      // http://www.netlib.org/lapack/explore-html/d9/d8e/group__double_g_eeigen_ga4f59e87e670a755b41cbdd7e97f36bea.html
+      else
+      {
+#ifdef PARANOID
+
+        // Are the eigenvalues finite?
+        if ((beta_eval[eval_count] != 0.0) &&
+            (beta_eval[eval_count + 1] != 0.0))
+        {
+          std::complex<double> lambda_this =
+            alpha_eval[eval_count] / beta_eval[eval_count];
+
+          std::complex<double> lambda_next =
+            alpha_eval[eval_count + 1] / beta_eval[eval_count + 1];
+
+          // Check cc-ness to within tolerance
+          if (fabs(lambda_this.imag() + lambda_next.imag()) >
+              Tolerance_for_ccness_check)
+          {
+            std::ostringstream error_stream;
+            error_stream << "Non-zero imaginary part of eigenvalue "
+                         << eval_count << " : " << lambda_this.imag()
+                         << std::endl;
+            error_stream
+              << "isn't the negative of its subsequent value       : "
+              << lambda_next.imag() << std::endl
+              << "Their sum " << (lambda_this.imag() + lambda_next.imag())
+              << " is greater than Tolerance_for_ccness_check = "
+              << Tolerance_for_ccness_check << std::endl;
+            throw OomphLibError(error_stream.str(),
+                                OOMPH_CURRENT_FUNCTION,
+                                OOMPH_EXCEPTION_LOCATION);
+          }
```

Comment:
throw separate exceptions or combine and throw one exception?

Reply:
Nothing wrong with what's done (I think!).

  Done.